src/data_source/fs.rs

```diff
@@ -0,0 +1,673 @@
+impl<Types: NodeType, I: NodeImplementation<Types>, UserData>
+    FileSystemDataSource<Types, I, UserData>
+where
+    Block<Types>: QueryableBlock,
+{
+    fn consensus_metrics(&self) -> Result<PrometheusMetrics, MetricsError> {
+        self.metrics.get_subgroup(["consensus"])
+    }
+}
+
+#[async_trait]
+impl<Types: NodeType, I: NodeImplementation<Types>, UserData> StatusDataSource
+    for FileSystemDataSource<Types, I, UserData>
+where
+    Block<Types>: QueryableBlock,
+    UserData: Sync,
+{
+    type Error = MetricsError;
+
+    async fn block_height(&self) -> Result<usize, Self::Error> {
+        Ok(self.leaf_storage.iter().len())
+    }
+
+    async fn mempool_info(&self) -> Result<MempoolQueryData, Self::Error> {
+        Ok(MempoolQueryData {
+            transaction_count: self
+                .consensus_metrics()?
+                .get_gauge("outstanding_transactions")?
+                .get() as u64,
+            memory_footprint: self
+                .consensus_metrics()?
+                .get_gauge("outstanding_transactions_memory_size")?
+                .get() as u64,
+        })
+    }
+
+    async fn success_rate(&self) -> Result<f64, Self::Error> {
+        let total_views = self.consensus_metrics()?.get_gauge("current_view")?.get() as f64;
+        // By definition, a successful view is any which committed a block.
+        Ok(self.block_height().await? as f64 / total_views)
+    }
+
+    async fn export_metrics(&self) -> Result<String, Self::Error> {
+        self.metrics.prometheus()
+    }
+}
+
+impl<Types: NodeType, I: NodeImplementation<Types>, UserData> UpdateStatusData
+    for FileSystemDataSource<Types, I, UserData>
+where
+    Block<Types>: QueryableBlock,
+{
+    fn metrics(&self) -> Box<dyn Metrics> {
+        Box::new(self.metrics.clone())
+    }
+}
+
+#[cfg(any(test, feature = "testing"))]
+mod impl_testable_data_source {
+    use super::*;
+    use crate::testing::mocks::{MockNodeImpl, MockTypes, TestableDataSource};
+    use tempdir::TempDir;
+
+    #[async_trait]
+    impl<UserData: Default + Send + Sync + 'static> TestableDataSource
+        for FileSystemDataSource<MockTypes, MockNodeImpl, UserData>
+    {
+        type TmpData = TempDir;
+
+        async fn create(node_id: usize) -> (Self, Self::TmpData) {
+            let dir = TempDir::new(&format!("file_system_data_source_{node_id}")).unwrap();
+            (Self::create(dir.path(), Default::default()).unwrap(), dir)
+        }
+
+        async fn commit_version(&mut self) {
+            Self::commit_version(self).await.unwrap();
+        }
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use super::super::data_source_tests;
+    use super::FileSystemDataSource;
+    use crate::testing::mocks::{MockNodeImpl, MockTypes};
+
+    // For some reason this is the only way to import the macro defined in another module of this
+    // crate.
+    use crate::*;
+
+    instantiate_data_source_tests!(FileSystemDataSource<MockTypes, MockNodeImpl, ()>);
```

Comment:
Silly question but where does `instantiate_data_source_tests` come from?

Reply:
It's defined by the [`#[generic_tests]`](https://github.com/EspressoSystems/hotshot-query-service/pull/251/files#diff-7136d6b9f26b655819ccf69416ccecb4d815aa64427a81900de400e450a44f15R32) macro